### PR TITLE
[PS-1452] Desktop/Browser: use readonly/disabled fields in cipher view, add copy buttons to all fields

### PR DESCRIPTION
**Pull request by patrickhlauke**
## Type of change

```
- [x] Bug fix
- [x] New feature development
- [x] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Currently, in both the desktop and browser versions, when viewing a cipher, most fields lack any underlying structure - the label and value are simply `<div>`s and `<span>`s. Structurally/semantically, this is a bit "loose". And for screen reader users, it means that they can only get to/hear the actual values by navigating using reading keys/cursor keys.

This PR switches the view to use `readonly` (and in the case of boolean custom fields, `disabled`, as you can't have a `readonly` checkbox currently) form fields with a proper `<label>`. This provides an unambiguous structure.

While initially in this PR I tried to provide some focus styling for when readonly fields receive focus, I've pivoted to a different approach: making the readonly fields explicitly non-focusable using `tabindex="-1"`.

To improve the keyboard experience, this PR now also adds explicit copy buttons for *all* fields (including `<textarea>` for notes/addresses). For screen reader users, the copy buttons now also refer back to their original field / value using `aria-describedby`, making for a much more immediate experience; in the case of secret/hidden fields, the value is only announced as part of the accessible description *if* the field value is also visible, providing the same level of privacy consideration.

The only fields that still don't receive focus and don't have a copy button are custom field checkboxes and linked fields, but for these it makes no sense to include a copy button.

## Code changes

- **apps/browser/src/_locales/en/messages.json**, **apps/desktop/src/locales/en/messages.json**, **apps/browser/src/_locales/en/messages.json**: add new strings for the various specific copy buttons
- **apps/browser/src/popup/scss/box.scss** and **apps/desktop/src/scss/box.scss**: added a new class `.box-content-row-textarea` for the textarea containers, and a new `.action-buttons-absolute` class to position the copy button for textareas absolutely *over* the corner of the textarea, rather than awkwardly off to the right and making the `<textarea>` smaller
- **apps/web/src/scss/forms.scss**: added new class `.input-group-append-absolute` and tweaked `.input-group` to allow for textarea/notes copy button to be absolutely positioned *over* the corner of the textarea
- **apps/browser/src/popup/vault/view.component.html:** Remove redundant `aria-readonly="true"` if there is already a `readonly` attribute ... the `aria` one is superfluous
- **apps/desktop/src/app/vault/view.component.html**, **apps/desktop/src/app/vault/view-custom-fields.component.html**, **apps/browser/src/popup/vault/view.component.html**, **apps/browser/src/popup/vault/view-custom-fields.component.html**: change the generic `<div>`/`<span>`s used to show a fake label and value to use actual `<label>` and `<input>`/`<textarea>` elements, but set to `tabindex="-1"` so they're not in focus order. Added "Copy..." buttons to all fields. Tweaked/fixed some of the existing `aria-label` definitions. Tied action buttons explicitly to their field/value using `aria-describedby` so that the value is announced by screen readers when the action button receives focus (with extra handling for password/hidden fields to only do this when the field is visually presented in clear text). As an aside, this also harmonises the desktop markup to be more in line with the the popup one - moving towards what I assume would be a desirable ultimate goal of having the exact same code for both? The change here also fixes a small inconsistency: copy buttons for websites incorrectly announce as "Copy URI" at the moment - now, website ones correctly announce as "Copy Website".
- **apps/web/src/app/vault/add-edit.component.html**: add copy buttons to all item view/edit fields in web vault

Edit: Split out the change originally made to `::selection` into its own PR https://github.com/bitwarden/clients/pull/3675. Backed out the change to the `copyVerificationCodeTotp` string on desktop, as it would cause problems with already translated versions (per https://github.com/bitwarden/clients/pull/3485#issuecomment-1265888668)

## Screenshots

Video recording on https://www.youtube.com/watch?v=xLTCkh12Er4 (too big for github inline video)

[video thumbnail](https://www.youtube.com/watch?v=xLTCkh12Er4)

Video shows the browser extension before/after in Windows/Chrome/NVDA and Windows/NVDA/desktop app. In both cases, we navigate using Tab/Shift+Tab (with a tiny digression at 0:47 - 1:00 to show how screen reader users *can* still get to the values using cursor keys/reading keys, but it's cumbersome)

---

Current desktop app showing an identity item (which currently is the one with the least/no copy buttons or anything)

image

After PR applied, desktop app showing an identity item - now with added copy buttons

image

---

Comparison of current and modified browser extension

A card item (note also the copy button on the textarea)

image

As above for desktop, comparion of an identity item (now with added copy buttons)

image

---

After screenshot of an item view in the web vault

image

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

**Comment by patrickhlauke** [comment]
Note that depending on when things get hopefully merged, this will need to be harmonised with https://github.com/bitwarden/clients/pull/3321, https://github.com/bitwarden/clients/pull/2660, and https://github.com/bitwarden/clients/pull/3347

**Comment by danielleflinn** [comment]
Hi @patrickhlauke! Thanks for another great accessibility contribution :) 

Can you provide more context around the WCAG requirement for a visible focus indicator? I'm assuming this relates to WCAG SC 2.4.7. Does the existing View Item focus state (highlight text) not meet the focus visible requirement? 
cc: @bitwarden/dept-design

**Comment by patrickhlauke** [comment]
@danielleflinn yes, it's 2.4.7. looking at the current situation (let's just take the browser popup as an example), it's fine except for the `<textarea>` (e.g. a note) in the view - it takes focus, but has no focus indication. so that needed to be addressed. and then now with my change, since there's more things that receive focus (like the TOTP and the password / "hidden" fields), and they wouldn't show the highlight (since for some of these I visually hide the actual control that receives focus, so that it can still use the custom colourised version for the password when revealed, which you couldn't have if just using `<input type="text">`) ... so to sidestep the issue, decided to give a more consistent and explicit focus style for all these extra new scenarios (and the old case of the `<textarea>`)

**Comment by patrickhlauke** [comment]
Following https://github.com/bitwarden/clients/pull/3321 being merged, I've updated this PR to incorporate the changes from there (making the `<label>` elements draggable)

**Comment by danielleflinn** [comment]
Thanks for the explanation @patrickhlauke . We definitely should provide a visible focus indicator for the fields you mentioned above that do not have the highlight. Rather than introducing a new focus UI pattern, could you use UI more consistent with existing patterns? Either the fake text-highlight you proposed or something similar to the focus outline that is used for the icon buttons should work (Design has a preference for the former).

**Comment by danielleflinn** [comment]
@patrickhlauke can you also provide a screen recording of the revised chrome/nvda screen reader functionality similar to the one you uploaded showing current screen reader functionality?

**Comment by patrickhlauke** [comment]
> @patrickhlauke can you also provide a screen recording of the revised chrome/nvda screen reader functionality similar to the one you uploaded showing current screen reader functionality?

the video in the OP shows the before and (at about 1:00) after (but only for the popup, desktop app is pretty much the same)

**Comment by patrickhlauke** [comment]
> Thanks for the explanation @patrickhlauke . We definitely should provide a visible focus indicator for the fields you mentioned above that do not have the highlight. Rather than introducing a new focus UI pattern, could you use UI more consistent with existing patterns? Either the fake text-highlight you proposed or something similar to the focus outline that is used for the icon buttons should work (Design has a preference for the former).

i'll have a look. not 100% i can fake text highlighting, as i don't think i can force the browser-native highlighting, so would need to pick some foreground and background colour explicitly. will have a try anyway.

**Comment by jtouchstonebw** [comment]
The design team has discussed the following options and is ok with either of these options for read-only focus states. Whichever one is easier to implement will work for us. 
Extension- View item Read only focus state 1@2
![Extension- View item Read only focus state 2@2x
x](https://user-images.githubusercontent.com/95641831/191091182-5e986885-19ba-4636-98df-76675cb5c40c.jpg)

**Comment by patrickhlauke** [comment]
a heads-up that i'm actually taking some time to make a more radical change proposal here:

* making the `readonly` fields also non-focusable with `tabindex="-1"`
* making sure that all fields have an equivalent copy action button (including `<textarea>`, with a copy button floating in the top-right of the area itself)

this makes the keyboard experience more consistent (compared to current state, where some fields are already readonly inputs, and are focusable despite having an explicit copy button, while others are just static text with a copy button) and cleaner (no need for duplicate focus stops - once on the readonly input, and then on the copy button)

I should have something by end of this weekend for consideration

**Comment by patrickhlauke** [comment]
@danielleflinn as mentioned, I completely changed approach/took it further with this PR - updated the opening comment to explain the changes.

**Comment by patrickhlauke** [comment]
additional point: for the changes to the json string files, I only made changes to the `en` version. Please let me know if changes (such as the renaming of the `copyVerificationCodeTotp` string for desktop to `copyVerificationCode`, should also be made as part of the PR, and I can include them. Also happy to include `en-gb` variant and `en-IN` if needed.

**Comment by patrickhlauke** [comment]
> PS: Only need to add/remove entries from the en/messages.json. This is our master/source for all translations. The rest is done in Crowdin and then pulled back into the clients (~1-2 weeks)

cool, thanks for confirming (I knew it was the case for new strings, but wasn't sure due to that renamed string, but as i'm backing it out now anyway the issue is moot

**Comment by patrickhlauke** [comment]
@djsmith85 

> Rename of `copyVerificationCodeTotp`

reverted that specific change from this PR. as said, for the other stuff, not sure it'd make sense to split over separate PRs as the new copy buttons do rely on the new strings.

Can, hover, back out the `::selection` change and move to a standalone PR

**Comment by patrickhlauke** [comment]
@djsmith85 

> Can, hover, back out the `::selection` change and move to a standalone PR

Done - see https://github.com/bitwarden/clients/pull/3675

**Comment by djsmith85** [comment]
@patrickhlauke Thank you.

I meant: to extract the buttons with their accompanying strings/messages into a separate PR, keeping the other changes here. I don't see the strings needed for anything else than the buttons (example: copyCardholderName)

**Comment by patrickhlauke** [comment]
oh, i also completely misread your original comment Daniel. i somehow read it as splitting the copy buttons into one PR, and then the messages change into another PR.

so making this just a change to readonly fields.

i guess this *could* be done, but would feel a bit icky as it just changes the underlying semantics without fixing the issue of keyboard/AT users (assuming i kept them non-focusable). or, if kept focusable, it'd go back to introducing an issue of a lack of focus indication for non-input fields (like passwords/secret fields/textareas), and i wouldn't want to initially push a PR that makes things worse from an #a11y point of view

**Comment by patrickhlauke** [comment]
if it's ok (though understand not ideal/easy to deal with), i'd like to keep this as a unit and deal with any merge conflicts etc as they arise, unless there's an explicit aspect that would prevent this/stop this from being considered

**Comment by djsmith85** [comment]
Thanks Patrick, thats a valid concern, breaking things again if those two potential PRs do not get merged at the same time/make the cut for a release.

I'll leave this up to @bitwarden/dept-design to decide. Once they are good with the changes this will get passed on to a developer review. 

I was only concerned about the scope of the changes in one PR.... Watching out for possibly, my future self 😉 😂

**Comment by patrickhlauke** [comment]
also, some foreshadowing: as a possible follow-up to this, once all "rows" have at least a "Copy" button, it's then worth considering only showing the action buttons when focus is within a row (`:focus-within`) and/or the mouse is hovering over it. and as a bonus, then adding a touch/mouse handler (not needed for keyboard) that also fires "Copy" as the default action whenever a user clicks/taps anywhere else on the row...similar to what a particular other password manager does...

two rows - username and password - with one hovered. the hovered one shows a "Copy" option, the other one doesn't

but that's a topic/suggestion for another day. however, this PR would pave the way towards this, if that's a direction you're wanting to go.

**Review by danielleflinn — CHANGES_REQUESTED** [review]
Design is good with the changes outlined; the team especially appreciates the addition of the copy button on all fields. Please update the new strings to use sentence casing; once this is done, we'll get it approved and reviewed by engineering.

File: apps/desktop/src/locales/en/messages.json [hunk 1]
```diff
@@ -138,42 +138,75 @@
   "cardholderName": {
     "message": "Cardholder name"
   },
+  "copyCardholderName": {
+    "message": "Copy Cardholder Name"
```
[the remark this answers is not in the document]
Comment (patrickhlauke) [reply apps/desktop/src/locales/en/messages.json]: @danielleflinn done

[the remark this answers is not in the document]
Comment (patrickhlauke) [reply apps/desktop/src/locales/en/messages.json]: oh sorry, i stupidly just did that one, but you meant all of them of course. give me a minute :)

[the remark this answers is not in the document]
Comment (patrickhlauke) [reply apps/desktop/src/locales/en/messages.json]: @danielleflinn there...all done properly this time. not being in the US, I still assume that "Social Security Number" is remains capitalised?

[the remark this answers is not in the document]
Comment (danielleflinn) [reply apps/desktop/src/locales/en/messages.json]: Thanks! We actually are using "Social Security number" as it appears on the US government website. Thanks for checking!

**Comment by djsmith85** [comment]
@patrickhlauke I know this is a long PR, with discussion and splitting things out over time. 

We love the addition of the copy buttons and would like to ask if you'd be up to adding them, also to the vault-items (login, card, identity) in the web-vault.

Either way you decide, we really appreciate the work you have done so far, and we'll obviously wrap-up the review here.

**Comment by djsmith85** [comment]
> @djsmith85 sorry, my email's been out of action the last two days, so not seen notifications and only just stumbled across this comment (and comments you left on the other two PRs). I'll grab some time this week to address this (don't think I ever tried running the web version locally for testing, so that will be exciting...)

@patrickhlauke No worries and thank you, Here's a [link](https://contributing.bitwarden.com/getting-started/clients/web-vault/)  to our setup guide to start up the web vault. Let me know if that works for you.

**Comment by patrickhlauke** [comment]
@djsmith85 perhaps a silly question, but as I don't normally use the web vault: is there no straight "View" view for entries? I only ever see the "Edit" view right away when choosing an item

screenshot of the edit dialog for a test card item

**Comment by patrickhlauke** [comment]
@djsmith85 done ... though admittedly it starts to look a bit busy, particularly because there is no distinction between viewing and editing items in the web vault - this makes keyboard navigation itself possibly more tedious, as now there's always two focus stops per field.

screenshot of an identity item in the web vault, now with copy buttons on all fields

taking a step back, is there any plan to actually make the web vault the same/similar to the desktop version?

**Review by danielleflinn — APPROVED** [review]

**Comment by patrickhlauke** [comment]
@danielleflinn 

> For the notes field copy button, can we move the copy button so it is next to the field label (similar to the "check password" action on the Password input)?

Do you mean just for the web vault, or the desktop/browser as well? And, if you meant just for the web one, note that I tweaked this further from that last screenshot (I've now added this to the OP) to match the approach I took on desktop and browser

image

I think this looks quite slick, but let me know if it needs changing (and if you meant it for desktop/browser as well?)

**Comment by danielleflinn** [comment]
This works; thanks for providing the updated screenshot; that said it looks like the right top/bottom corners of the notes field have lost their roundness in web vault? Can you double check the corner radius is still being applied?

**Comment by patrickhlauke** [comment]
@danielleflinn ah, well spotted. it's because that copy button (with hard corners) is actually on top of the textarea. I can apply an extra patch to make *that* specific copy button have a top-right rounded corner though (this wasn't an issue on desktop/browser because there the copy buttons don't have their own button-y stylings)

**Comment by patrickhlauke** [comment]
@danielleflinn fixed

screenshot with zoomed-in notes textarea, showing rounded corners all around

**Review by danielleflinn — DISMISSED** [review]
Looks good thanks Patrick!

**Comment by patrickhlauke** [comment]
@danielleflinn once this is merged, I may have a crack at the new functionality/behaviour I hinted at ... having those copy buttons only visually appear when mouse is hovering over the field / keyboard focus is on the button (provided you think that's a direction the design team would be happy to go for)

**Comment by danielleflinn** [comment]
@patrickhlauke you can hold off on only showing the copy buttons on hover. Design avoids only showing actions on hover to improve discoverability of features.

**Comment by patrickhlauke** [comment]
@Hinton a few aspects in your question that could probably do with a much longer explanation, but I'll try to summarise it ... happy to chat more in depth about it though:

* there is no standard construct for marking up static key/value pairs in a semantic/structural way. the closest you come to it is using a `<table>` with proper `<th>`/`<td>` relationships (which is overkill  for individual key/value pairs, and very inflexible if you want to do it for many key/value pairs without also having a table-like visual presentation) or a description list `<dl>` (with key in `<dt>` and value in `<dd>`) - though the latter is exposed in a strange way in most screen readers.
* using `aria-label`, `aria-labelledby`, `aria-describedby` on static elements without a semantic role (e.g. a regular `<span>` or `<div>`) does not work consistently ... you need to "hang" those on elements that already have a (intrinsic, by virtue of their element itself, or explicit, using `role="..."`) role that expects to have a label/description. see https://www.tpgi.com/short-note-on-aria-label-aria-labelledby-and-aria-describedby/ - so there's also no good/natural way to use ARIA to mark up a key/value pair
* in terms of abusing/perverting HTML elements, because the item view is essentially the read-only version of the edit view, and you can flip between them, it feels less like an abuse of HTML as such. if this was purely a standalone page (with no "edit" view related to it), i'd go for something like the `<dl>` or `<table>` approach, but because there's already the precedent of having the proper inputs when editing, it feels ok (to me) to say that the view itself uses readonly fields.
* "Inputs are quite a bit more restrictive in terms of how we can visually display them. A good example of this is the card numbers" indeed, inputs are restrictive ... but for those cases (see also passwords), note that I actually used a visually hidden input (with the `sr-only` class) and kept the visible, nicely formatted/coloured/etc text (but made that `aria-hidden="true"` so it's not double-announced by screen readers) - see for instance this https://github.com/bitwarden/clients/pull/3485/files#diff-aa27c4cb976574a37f6d42b3e54f39215545c6876a070d3dbf85c3b9bf6db947R30 - so i don't think there's a risk of "degraded experience and/or capability for future changes"
* "Some browser also seem to persist form values in memory longer than non form values, possibly since forms sometimes persist when navigating backwards." not sure what you mean here...you saying that in some browsers, even if the underlying value changed (e.g. the item was edited), it will show the old value? not something i've encountered, and it shouldn't really happen?
* "My other concern is that we are planning on migrating our other clients to the component library after we've completely removed bootstrap from the web vault. This will necessitate large changes to the application and unless we write up guidelines for using inputs (which restricts what we can visually display) I suspect these changes won't be re-applied." as mentioned above, there should not be any restrictions here, since for the values that have special display, the real `<input>` is visually hidden, and the formatted/styled text is there as before, just set to `aria-hidden="true"`

**Comment by patrickhlauke** [comment]
I'll need a bit of time to resolve the merge conflicts. hoping to carve out some time tomorrow.

**Comment by patrickhlauke** [comment]
merge conflicts addressed

**Comment by Hinton** [comment]
What about using something like https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/rowheader_role?

That would let us simplify the required changes to just adding roles to the container, label and value? I haven't tested how well supported those wai-roles are though.

**Comment by patrickhlauke** [comment]
> What about using something like https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/rowheader_role

you can't just use that in isolation. at that point, you need to effectively build an ARIA table/grid. doable, sure, but then you need to decide if you want to make lots of "mini-tables" for each key/value pair (making it extremely verbose in terms of screen reader output, as it will keep announcing that you're in a table with 1 column and 2 rows or similar) or turn the whole section/view into a giant single grid/table using ARIA. you'd then also need to remove all of that ARIA in edit mode.

**Comment by patrickhlauke** [comment]
it would probably be good to get a go/no-go decision on this at some point. just to reiterate that the "label with a read-only input when viewing, turning into a label with editable input when editing" approach is, in my view, not as exotic/uncommon - see for instance https://getbootstrap.com/docs/5.3/forms/form-control/#readonly-plain-text (though in bitwarden's case, we're additionally making them inputs non-focusable by default with `tabindex="-1"` because we're also providing custom "Copy" buttons, which makes the ability to set focus to the readonly inputs for copy/pasting purposes when navigating with a keyboard redundant)

**Comment by patrickhlauke** [comment]
@Hinton @danielleflinn any more thoughts on this?

**Comment by patrickhlauke** [comment]
@Hinton @djsmith85 should I work on fixing the merge conflicts, or abandon this PR?

**Comment by patrickhlauke** [comment]
doing a merge on desktop (rather than relying on github's automatic merge/update) seems to have worked fine, without needing to do any manual conflict resolution (unless i somehow managed to mess things up locally)

**Comment by patrickhlauke** [comment]
i don't want to constantly push on this, but at this point i'm getting concerned that this PR will start going out of sync with other stuff being merged. luckily there's only been two recent cases where some merge conflicts were flagged by github, though then merging locally on desktop seemed to resolve things without the need for any intervention. but it would be good to know what's preventing this currently from being merged. is it just too much of a change? too many concerns with the use of `readme` inputs?

**Comment by patrickhlauke** [comment]
@djsmith85 @danielleflinn @Hinton ping?

**Comment by patrickhlauke** [comment]
@djsmith85 @Hinton  is it worth me resolving the merge conflicts, or should I just close this PR now? any indication would be appreciated...

**Comment by patrickhlauke** [comment]
as we're coming up to 11 months that this PR has been left here, and the merge conflicts are now just piling up as a result... going to close it. shame that this didn't even deign a response though folks...not impressed.

**Pull request closed without merging**